instrgen/docs/design.md

```diff
@@ -0,0 +1,186 @@
+# OpenTelemetry Go Source Automatic Instrumentation - design
+
+This document describes design of `instrgen` - compile time automatic instrumentation for GO programming language.
+
+## Background
+
+Usually techniques used for instrumenting programs in specific languages are connected to runtimes and type systems
+used by these languages and depend on capabilities offered by them. Methods differs depending on language properties.
+Is executed by virtual machine? Is statically or dynamically typed?
+What's matter is the injection point.
```

Comment:
```suggestion
Are they executed by virtual machine? Are they statically or dynamically typed?
What are the injection points.
```

Reply:
I'm thinking about last sentence. What I meant here is that language internals influence instrumentation technique.